oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchFacetTest.java

```diff
@@ -0,0 +1,466 @@
+        session.save();
+
+        createDataset(NUM_LEAF_NODES_FOR_LARGE_DATASET);
+
+        assertEventually(() -> {
+            assertEquals("Unexpected number of facets", actualAclLabelCount.size(), getFacets().size());
+        });
+
+        for (Map.Entry<String, Integer> facet : actualAclLabelCount.entrySet()) {
+            String facetLabel = facet.getKey();
+            assertEventually(() -> {
+                int facetCount = facetCount = getFacets().get(facetLabel);
+                float ratio = ((float) facetCount) / facet.getValue();
+                assertTrue("Facet count for label: " + facetLabel + " is outside of 10% margin of error. " +
+                                "Expected: " + facet.getValue() + "; Got: " + facetCount + "; Ratio: " + ratio,
+                        Math.abs(ratio - 1) < 0.1);
+            });
+        }
+    }
+
+    @Test
+    public void statisticalFacetsWithHitCountLessThanSampleSize() throws Exception {
+        Node facetConfig = getOrCreateByPath(indexNode.getPath() + "/" + FACETS, "nt:unstructured", session);
+        facetConfig.setProperty(PROP_SECURE_FACETS, PROP_SECURE_FACETS_VALUE_STATISTICAL);
+        indexNode.setProperty(PROP_REFRESH_DEFN, true);
+        session.save();
+
+        createDataset(NUM_LEAF_NODES_FOR_SMALL_DATASET);
+
+        assertEventually(() -> {
+            assertEquals("Unexpected number of facets", actualAclLabelCount.size(), getFacets().size());
+        });
+
+        // Since the hit count is less than sample size -> flow should have switched to secure facet count instead of statistical
+        // and thus the count should be exactly equal
+        assertEventually(() -> {
+            assertEquals(actualAclLabelCount, getFacets());
+        });
+    }
+
+    /*
+    Currently we are not adding path restrictions in elastic query and filtering paths later from results.
+    We will need path restrictions in elastic query itself to get right facet results.
+     */
+    @Test
+    public void statisticalFacets_withHitCountSameAsSampleSize() throws Exception {
+        Node facetConfig = getOrCreateByPath(indexNode.getPath() + "/" + FACETS, "nt:unstructured", session);
+        facetConfig.setProperty(PROP_SECURE_FACETS, PROP_SECURE_FACETS_VALUE_STATISTICAL);
+        indexNode.setProperty(PROP_REFRESH_DEFN, true);
+        session.save();
+
+        createDataset(NUM_LEAF_NODES_FOR_LARGE_DATASET);
+
+        Map<String, Integer> facets = getFacets("/parent/par1");
+        assertEquals("Unexpected number of facets", actualAclPar1LabelCount.size(), facets.size());
+
+        for (Map.Entry<String, Integer> facet : actualAclPar1LabelCount.entrySet()) {
+            String facetLabel = facet.getKey();
+            int facetCount = facets.get(facetLabel);
+            float ratio = ((float) facetCount) / facet.getValue();
+            assertTrue("Facet count for label: " + facetLabel + " is outside of 10% margin of error. " +
```

Comment:
The test case uses a random generator and it may happen sometimes that the facetLabels have too skewed count leading to error you mentioned.
Is this one of the case or this happened only once?

Reply:
I run it twice and I got the same error result both times.

Reply:
Looks like problem with different java... oracle vs openJDK.